### PR TITLE
cmake: fix Debug build `WITH_SEASTAR=ON`

### PR DESCRIPTION
**Pull request by tchaikov**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/osd/CMakeLists.txt [hunk 1]
```diff
@@ -44,8 +44,9 @@ add_library(osd STATIC ${osd_srcs}
   $<TARGET_OBJECTS:cls_references_objs>
   $<TARGET_OBJECTS:global_common_objs>)
 target_link_libraries(osd PRIVATE
+  os dmclock heap_profiler cpu_profiler
   ${LEVELDB_LIBRARIES}
-  dmclock heap_profiler cpu_profiler ${CMAKE_DL_LIBS})
+  ${CMAKE_DL_LIBS})
```
Comment (cbodley) [inline src/osd/CMakeLists.txt]: i see that we still link os in `target_link_libraries(ceph-osd osd os`, can we make this one PUBLIC so it isn't needed in ceph-osd?
```cmake
target_link_libraries(osd
  PUBLIC os
  PRIVATE dmclock heap_profiler cpu_profiler
```

**Comment by tchaikov** [comment]
@cbodley updated and repushed.

**Comment by cbodley** [comment]
jenkins test make check

**Comment by wjwithagen** [comment]
@tchaikov 
I tried this in FreeBSD/Jenkins, and I got about 6-7 tests that failed.
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/2443/console

Removing the PR got things working again.
But I have no idea yet where the error is.
Rather busy with $work atm.

**Comment by tchaikov** [comment]
@wjwithagen i can hardly tell what's going wrong with the failures. most of them are either segfaults or bus errors. will setup a FreeBSD vm to reproduce these failures. at the meanwhile, i'm inclined to merge it after running it thru rados qa suite.

my wild guess is that the way how gtest is built in this change has a bad effect on the tests..

**Comment by wjwithagen** [comment]
@tchaikov 
I can live with that, but the set of changed files is substantial.
So figuring things out can be convoluted.
OTOH it could also be as simple as usign a set of different linking options, like one of the older PRs that did not work for FreeBSD.

**Comment by cbodley** [comment]
@tchaikov i'd prefer not to merge something that breaks freebsd

**Comment by tchaikov** [comment]
i tested this PR on FreeBSD 11.2-RELEASE, with clang 6. the tests passed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -585,8 +585,11 @@ endif()
 
 # Rocksdb
 option(WITH_SYSTEM_ROCKSDB "require and build with system rocksdb" OFF)
-if (WITH_SYSTEM_ROCKSDB)
+if(WITH_SYSTEM_ROCKSDB)
   find_package(RocksDB 5.8 REQUIRED)
+else()
+  include(BuildRocksDB)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply CMakeLists.txt]: please refer to the commit message.

File: cmake/modules/BuildGoogleTest.cmake [hunk 3]
```diff
@@ -0,0 +1,35 @@
+function(build_google_test gtest_root)
+  include(ExternalProject)
+  ExternalProject_Add(googlemock-ext
+    SOURCE_DIR ${gtest_root}
+    CMAKE_ARGS
+      -DBUILD_GMOCK=ON
+      -DBUILD_GTEST=OFF
+      -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+    INSTALL_COMMAND "")
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply cmake/modules/BuildGoogleTest.cmake]: i dropped this file from the latest changeset.

File: cmake/modules/FindSanitizers.cmake [hunk 4]
```diff
@@ -1,6 +1,6 @@
 if(NOT Sanitizers_FIND_COMPONENTS)
   set(Sanitizers_FIND_COMPONENTS
-    address undefined-behavior)
+    address undefined_behavior)
 endif()
 if(HAVE_JEMALLOC)
   message(WARNING "JeMalloc does not work well with sanitizers")
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply cmake/modules/FindSanitizers.cmake]: i am aware of this. and could you be more specific?

**Comment by tchaikov** [comment]
instead of building gtest/gmock as an external project, in the latest changset, they are built as sub-project as before. hopefully this would address the test failures on FreeBSD 12.0 CURRENT + clang 6.0.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (wjwithagen) [reply CMakeLists.txt]: @tchaikov 
I was refering to feb06ad: cmake: fix link dependencies 
That is all about cleanup, not linking gtest.
But I don't feel too strong about that, it just bloats the PR.

**Comment by wjwithagen** [comment]
@tchaikov 
I will run that new version thru Jenkins once more.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply CMakeLists.txt]: again, as i put in the commit message. it's not a cleanup.

**Comment by wjwithagen** [comment]
@tchaikov 
Rerunning Jenkins, since it caught the same error you just fixed.

**Comment by wjwithagen** [comment]
@tchaikov 
Nope, still errors with the PR:
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/2453/console

And if the PR is not applied, tests are oke:
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/2454/consoleFull

Maybe I have some time tonight to disect the PR and see where tings go wrong.
Could also be I need to upgrade 12.0-CURRENT, since it is 2 months old, and it is nearing its release.
But that is a bit more work.

**Comment by wjwithagen** [comment]
@tchaikov 
But I have the idea that the errors occur between the tests.
And that leads me to believe that the errors actually originate in the gtest code that surrounds the ceph-tests.

**Comment by tchaikov** [comment]
@wjwithagen thanks for testing. the generated debug symbol package of my test batch containing this PR is over 2G. the debug symbol package should be around 1.55G without this change. i will also look into this issue.

**Comment by wjwithagen** [comment]
@tchaikov 
When you tested on 11.2 what linker did you use?
Reason I ask: I'm using the new linker that comes with llvm, because with some unitttests it runs into a linking error. 
But if you have not changed anything, then that could be a reason for my errors.
Only question is how you sneaked by the build error I run into.
Will try with the std(gnu) linker.

**Comment by tchaikov** [comment]
@wjwithagen i didn't change any settings in specific. i just 

1. install `git` and `bash`
1. updated the ports using portsnap, and installed 
1. ran `install-deps.sh`
1. ran `do_freebsd.sh`.

but the vm was destroyed, so i can hardly tell what exactly linker i was using. if you are interested, i will try to create another this week-end.

**Comment by tchaikov** [comment]
@wjwithagen i am moving the gtest change into #23628. the size of generated debug-info package looks sane: https://shaman.ceph.com/builds/ceph/wip-cmake-gtest/6f231ee8640a332b3a04808b7a5e9126369ebe57/ . could you give it a try ?

**Comment by tchaikov** [comment]
now that the gtest change has been merged in #23628, i am removing it from this PR.

**Comment by wjwithagen** [comment]
@tchaikov 
Dust has setled more or less on the FreeBSD master builds, so I'll run this thru.

**Comment by tchaikov** [comment]
@wjwithagen 7d46e180957158b8a67d8ac2c3b34751c6047c02 and 3e5c77a70073a4eba8f3e442d0d3e779d1c73110 are fishy.  with these changes, the generated debug-info sizes > 2G. so i am dropping them.

**Comment by wjwithagen** [comment]
@tchaikov 
Jenkins seems to be happy:
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/2476/consoleFull
The refusal is the already merged other PR from this bunch, so that can be ignored.

**Pull request merged**